### PR TITLE
[stdlib] Adopt conditional conformance for Indices, Slice, ReversedCollection

### PR DESCRIPTION
**Pull request by airspeedswift**
This PR refactors the variations of `Slice`, `DefaultIndices`, and `ReversedCollection` to be conditional conformances on each base type e.g.

```swift
ReversedRandomAccessCollection<Base: RandomAccessCollection>: RandomAccessCollection
```

becomes:

```swift
extension ReversedCollection: RandomAccessCollection where Base: RandomAccessCollection { }

@available(*, deprecated, renamed: "ReversedCollection")
public typealias ReversedRandomAccessCollection<T: RandomAccessCollection> = ReversedCollection<T>
```

### CODE REVIEW

**Comment by slavapestov** [comment]
Wow that's a lot of deleted code.

File: stdlib/public/core/Indices.swift [hunk 1]
```diff
@@ -135,15 +119,16 @@ where Indices == ${Self}<Self> {
...
+  public var indices: DefaultIndices<Self> {
+    return DefaultIndices(
       _elements: self,
       startIndex: self.startIndex,
       endIndex: self.endIndex)
   }
 }
 
-% end
+public typealias DefaultBidirectionalIndices<T: BidirectionalCollection> = DefaultIndices<T>
```
Comment (moiseev) [inline stdlib/public/core/Indices.swift]: Deprecated in swift 4.1, perhaps?

File: stdlib/public/core/RangeReplaceableCollection.swift [hunk 2]
```diff
@@ -1,4 +1,4 @@
-//===--- RangeReplaceableCollection.swift.gyb -----------------*- swift -*-===//
+//===--- RangeReplaceableCollection.swift -----------------*- swift -*-===//
```
Comment (moiseev) [inline stdlib/public/core/RangeReplaceableCollection.swift]: Notpick: No longer 80 columns.

File: stdlib/public/core/Reverse.swift [hunk 3]
```diff
@@ -269,177 +269,11 @@ public struct ReversedCollection<
...
-  public subscript(position: Index) -> Base.Element {
-    return _base[_base.index(before: position.base)]
-  }
-
-  // FIXME: swift-3-indexing-model: the rest of methods.
-
-  public let _base: Base
-}
+@available(*, deprecated, renamed: "ReversedCollection")
+typealias ReversedRandomAccessCollection<T> = ReversedCollection<T> where T: RandomAccessCollection
```
Comment (moiseev) [inline stdlib/public/core/Reverse.swift]: `public`

File: stdlib/public/core/Indices.swift [hunk 4]
```diff
@@ -115,8 +84,23 @@ public struct ${Self}<
...
+  }
+
+  @_inlineable
+  public func formIndex(before i: inout Index) {
+    // FIXME: swift-3-indexing-model: range check.
+    _elements.formIndex(before: &i)
+  }
+}
+
+extension DefaultIndices: RandomAccessCollection where Elements: RandomAccessCollection { }
```
Comment (natecook1000) [inline stdlib/public/core/Indices.swift]: Need to make sure we're forwarding distance(to:) and index(_:offsetBy:) to the underlying collection.

File: stdlib/public/core/Reverse.swift [hunk 5]
```diff
@@ -269,177 +269,11 @@ public struct ReversedCollection<
...
-  @_inlineable
-  public var startIndex: Index {
-    return ReversedRandomAccessIndex(_base.endIndex)
-  }
-
-  @_inlineable
-  public var endIndex: Index {
-    return ReversedRandomAccessIndex(_base.startIndex)
-  }
+extension ReversedCollection: RandomAccessCollection where Base: RandomAccessCollection { }
```
Comment (natecook1000) [inline stdlib/public/core/Reverse.swift]: Here too — need the index-moving methods.

**Review by natecook1000 — COMMENTED** [review]
This is so great! :clap::clap::clap:

**Comment by DougGregor** [comment]
The first compiler crash this hits is: https://bugs.swift.org/browse/SR-6466

File: stdlib/public/core/Reverse.swift [hunk 6]
```diff
@@ -523,22 +314,6 @@ extension LazyCollectionProtocol
...
-extension LazyCollectionProtocol
-  where
-  Self : RandomAccessCollection,
-  Elements : RandomAccessCollection {
-
-  /// Returns the elements of the collection in reverse order.
-  ///
-  /// - Complexity: O(1)
-  @_inlineable
-  public func reversed() -> LazyRandomAccessCollection<
```
Comment (DougGregor) [inline stdlib/public/core/Reverse.swift]: Whoops, we still need this, up until the point where the Lazy collections also use conditional conformances

**Comment by moiseev** [comment]
@swift-ci Please smoke test macOS platrform

**Comment by moiseev** [comment]
@swift-ci Please smoke test macOS platform

**Comment by DougGregor** [comment]
@airspeedswift , that last commit re-breaks the test. The gyb'd conformances *are* redundant, so we (correctly) "redundant conformance" errors. Why did you revert my change? Because it uncovered a source-compatibility issue?

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
Quite curious how this fares on the compatibility suite in its current state.

**Comment by DougGregor** [comment]
Note that *all* of the test suite failure other that the `Prototypes/Algorithms.swift.gyb` one mentioned about are in the `subscript(_: Range)/Set/semantics` test at https://github.com/apple/swift/blob/master/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb#L244-L295.

**Comment by DougGregor** [comment]
... and the crash appears to be in the `materializeForSet` for Slice's range subscript.

```
_T0s5SliceVss17MutableCollectionRzrlEAByxGs5RangeVy5IndexQzGcimytfU_ ---> closure #1 : () in (extension in Swift):Swift.Slice< where A: Swift.MutableCollection>.subscript.materializeForSet : (Swift.Range<A.Index>) -> Swift.Slice<A>
```

**Comment by DougGregor** [comment]
The Plank and R failures are the SwiftPM failure under discussion in https://github.com/apple/swift-package-manager/pull/1393, so this isn't introducing any source-compatibility regressions.

**Comment by DougGregor** [comment]
Turns out that we don't need that big pile of explicit typealiases in the standard library now that I've reinstated the associated type redeclarations in the `Collection` lattice.

**Comment by airspeedswift** [comment]
@DougGregor sorry, I thought that was reverting a change I made from before you restricted the feature to only stdlib use (the test was failing). Yeah, if that original code doesn't compile, it does suggest a source compatibility issue we'll need to work around.

**Comment by moiseev** [comment]
@swift-ci Please smoke test macOS platform

**Comment by DougGregor** [comment]
The source compatibility issue here is that the code in question introduced conformances for each of the `Revered*Collection` types to a single new protocol, e.g.,

```swift
protocol P { }
extension ReversedCollection : P { }
extension ReversedRandomAccessCollection : P { }
extension ReversedRangeReplaceableCollection : P { }
```

With this PR, those conformances are now redundant, because it's stating `ReversedCollection : P ` three times.

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by moiseev** [comment]
@swift-ci Please test macOS platform

**Comment by moiseev** [comment]
@swift-ci please smoke test compiler performance

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by DougGregor** [comment]
FWIW, I rebased on master, fixed up the conflicts, and removed the extra hack + workaround commits.

**Comment by DougGregor** [comment]
Fixed up the `Prototypes/Algorithms.swift.gyb test`. The "only" remaining failures are those due to https://github.com/apple/swift/blob/master/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb#L244-L295.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
The only significant casualty on the benchmarks seems to be CharIteration in Debug. @milseman possibly something to be aware of...

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please test compiler performance

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@airspeedswift merge it! merge it! :)

**Comment by natecook1000** [comment]
Did we figure out a plan for https://github.com/apple/swift/pull/12913#issuecomment-347975121 ? If that's part of our compatibility suite it'll surely be elsewhere in the wild—I could definitely see people giving conformance to a new protocol to multiple `Slice` types, for instance. Do we need to keep these types around in deprecated form to have compatibility with Swift 4.0?

**Comment by DougGregor** [comment]
@natecook1000, that source-compatibility issue was in our own standard library test suite. We haven't seen any such code in the wild yet. The core team discussed that specific source compatibility concern and decided that the benefits of this change vastly outweigh the cost of such breakage; if we see more widespread breakage, we'll consider additional mitigations in the compiler (e.g, ignoring redundant conformances in such cases)

**Comment by natecook1000** [comment]
Thanks @DougGregor, glad to know it's been discussed!

**Pull request merged**